pallets/randomness/src/vrf.rs

```diff
@@ -0,0 +1,126 @@
+		// logs if input uniqueness assumptions are violated (no reuse of vrf inputs)
+		if last_vrf_input.storage_root == input.storage_root
+			|| last_vrf_input.slot_number == input.slot_number
+		{
+			log::warn!(
+				"VRF on_initialize: storage root or slot number did not change between \
+			current and last block. Nimbus would've panicked if slot number did not change \
+			so probably storage root did not change."
+			);
+		}
+	}
+	<CurrentVrfInput<T>>::put(input);
+}
+
+/// Returns weight consumed in `on_initialize`
+pub(crate) fn set_output<T: Config>() -> Weight {
+	let input = <CurrentVrfInput<T>>::get().expect("VrfInput must be set to verify VrfOutput");
+	let mut block_author_vrf_id: Option<VrfId> = None;
+	let PreDigest {
+		vrf_output,
+		vrf_proof,
+	} = <frame_system::Pallet<T>>::digest()
+		.logs
+		.iter()
+		.filter_map(|s| s.as_pre_runtime())
+		.filter_map(|(id, mut data)| {
+			if id == VRF_ENGINE_ID {
+				PreDigest::decode(&mut data).ok()
+			} else {
+				if id == NIMBUS_ENGINE_ID {
+					let nimbus_id = NimbusId::decode(&mut data)
+						.expect("NimbusId encoded in pre-runtime digest must be valid");
+
+					block_author_vrf_id = Some(
+						T::VrfKeyLookup::lookup_keys(&nimbus_id)
+							.expect("No VRF Key Mapped to this NimbusId"),
+					);
+				}
+				None
+			}
+		})
+		.next()
+		.expect("VRF PreDigest must be decoded from the digests");
+	let block_author_vrf_id =
+		block_author_vrf_id.expect("VrfId encoded in pre-runtime digest must be valid");
+	let pubkey = schnorrkel::PublicKey::from_bytes(block_author_vrf_id.as_slice())
+		.expect("Expect VrfId to be valid schnorrkel public key");
+	let transcript = make_transcript::<T::Hash>(input.slot_number, input.storage_root);
+	// NOTE: this is verified by the client when importing the block, before
+	// execution. we don't run the verification again here to avoid slowing
+	// down the runtime.
+	// TODO: verify by the client when importing the block
+	debug_assert!(pubkey
+		.vrf_verify(transcript.clone(), &vrf_output, &vrf_proof)
+		.is_ok());
+	let vrf_output: Randomness = vrf_output
+		.attach_input_hash(&pubkey, transcript)
+		.ok()
+		.map(|inout| inout.make_bytes(&VRF_INOUT_CONTEXT))
+		.expect("VRF output encoded in pre-runtime digest must be valid");
+	let raw_randomness_output = T::Hash::decode(&mut &vrf_output[..]).ok();
+	if raw_randomness_output.is_none() {
+		log::warn!("Could not decode VRF output bytes into Hash Type");
+	}
+	LocalVrfOutput::<T>::put(raw_randomness_output);
+	// Supply randomness result
+	let local_vrf_this_block = RequestType::Local(frame_system::Pallet::<T>::block_number());
+	if let Some(mut results) = RandomnessResults::<T>::get(&local_vrf_this_block) {
+		if let Some(randomness) = raw_randomness_output {
+			results.randomness = Some(randomness);
+			RandomnessResults::<T>::insert(local_vrf_this_block, results);
+		} else {
+			log::warn!("Could not read local VRF randomness from the relay");
```

Comment:
What are the ramifications of hitting this (or 5 lines above if `get()` returns `None`)?